Add GUI with inputs for VMware vCloud event monitoring

miha-plesko opened the pull request:
User had to input her RabbitMQ credentials using Rails console.
Not any more - it can be done via GUI at ems cloud provider form (see screenshots below).

User can opt-in to avoid AMQP:
![event_gui_1](https://cloud.githubusercontent.com/assets/8102426/19350687/0b670c40-9159-11e6-9898-25c86685af6c.PNG)

Or she can provide AMQP credentials:
![event_gui_2](https://cloud.githubusercontent.com/assets/8102426/19350686/0b6093ba-9159-11e6-953b-7a0d77536eb8.PNG)

miha-plesko commented:
Hello guys, I'd kindly ask for some guides regarding angular/ruby support for this PR.
I've managed to render input forms as shown in the screenshots above, but no matter what I input as AMQP credentials, the same error appears:

![event_gui_2](https://cloud.githubusercontent.com/assets/8102426/19078480/8b4032e4-8a50-11e6-93e7-89ba1332f9c2.PNG)

My credentials are correct for sure, the problem is that in [this line](https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/vmware/cloud_manager/manager_events_mixin.rb#L10) I always get `nil` meaning that credentials are not parsed from HTML form. The `amqp` connection configuration role seems not to be created and I am not sure why...

Any idea what I'm doing wrong?

sergio-ocon commented:
Hi @dclarizio can you help with this? We don't seem to be able to make it work

h-kataria commented:
@AparnaKarve can you help out with this one

himdel commented:
Steps to reproduce.. go to Compute > Cloud > Providers .. add a provider..

Choose type VMware vCloud, and switch to Events endpoint, and change to AMQP.

Filling out the fields and clicking validate does send a POST to /ems_cloud, with the filled in values..

```
event_stream_selection:amqp
amqp_hostname:foo
amqp_api_port:5672
amqp_userid:bar
amqp_password:baz
amqp_verify:baz
```

So the error is very probably on the ruby side (unless the prefix is supposed to include Events).

himdel commented:
@miha-plesko `connection_configuration_by_role` looks at `endpoints` to find an endpoint with the given role ("amqp" this time). 

But `endpoints` only contains one endpoint, of type default.

So there's your problem, hope that helps ;)

AparnaKarve commented:
@sergio-ocon We know that in general, the credential validation works OK. Hence wanted to check with you if you were able to ping that hostname from a terminal?
The flash error message from the Validation: `Could not establish TCP connection` suggests that there seems to be an issue with the hostname.

miha-plesko commented:
Thank you @himdel. How come I didn't think of just go look at what POST request is sent :smile:
Anyway, I was aware that endpoints table does not contain amqp, but I am not sure who should put it there. Do you perhaps know what ruby file to look at? Is this table to be updated already at validation time?

himdel commented:
@miha-plesko first thing to try when in doubt ;)

Anyway, sorry, endpoint-related questions will need to go in the providers channel, no idea there..

gberginc commented:
@miha-plesko [this](https://github.com/ManageIQ/manageiq/pull/11682/files#diff-f402e411cd6dc158c724b14d72ead612R386) is what you were missing in your initial commit. I have also changed `event_monitor_options` slightly to be more in line with how OpenStack provider sets options, and rebased it.

AMQP credentials are validated now in the UI.

I haven't tested whether the updates are actually working. Please, make sure to test this, remove WIP and notify @agrare.

gberginc commented:
@h-kataria @himdel can you please review the changes to the UI?

AparnaKarve commented:
Quick question: The default tab currently does not have `Security Protocol`. 
Does that seem right for VMware vCloud?

gberginc commented on app/views/layouts/angular/_multi_auth_credentials.html.haml:
```diff
@@ -188,13 +188,59 @@
+                                    :id                     => record.id,
+                                    :prefix                 => "amqp",
+                                    :ng_reqd_api_port       => "false",
+                                    :ng_reqd_hostname       => "false"}
```
@miha-plesko Besides the comment from @AparnaKarve, I am interested in why you had to set `ng_reqd_hostname` to `false`? For port I understand it's not required as a default can be used in most cases, but hostname should not be a default, I guess...

miha-plesko replied:
True, thanks for noticing this!

gberginc commented on app/views/layouts/angular/_multi_auth_credentials.html.haml:
```diff
@@ -188,13 +188,59 @@
                                     :id                => record.id,
                                     :prefix            => "amqp",
                                     :basic_info_needed => true}
+
```
Do you think it would make sense to merge this part L191-231 with OpenStack? We now have three options (none, ceilometer and amqp) but the two sections are pretty much duplicated.

miha-plesko replied:
I've attempted to merge with OpenStack tab, but there were too much ng-if statements for my taste. I'd prefer to go with some duplication for code to remain readable/managable.

miha-plesko commented:
@gberginc thank you for adding the commit, it seems to be working for me. I'm applying the comments above then I'll update the PR.

miha-plesko commented on app/models/manageiq/providers/vmware/cloud_manager/event_catcher/stream.rb:
```diff
@@ -12,15 +12,11 @@ def self.test_amqp_connection(options = {})
-      _log.info("Failed testing rabbit amqp connection for #{options[:hostname]}. ")
       raise
-    ensure
-      connection.close if connection.respond_to? :close
```
What I'm doing here is I'm prettifying the message that user gets when she validates with incorrect hostname or username/password. The `ensure` part had to be removed since it timeout-ed attempting to close a not-even-opened connection. Then user was always given message about timeout even if the problem was actually incorrect password :grimacing:

gberginc replied:
Where do you close the connection when the test succeeds?

gberginc replied:
@miha-plesko I wonder if we can get into a state where this wouldn't work properly, namely an exception would occur, but the connection would be open somehow. Would this additional condition in the deleted line 23 help resolving your case: 

```
connection.close if connection.respond_to? :close and connection.open?
```

From [Bunny doc](http://www.rubydoc.info/github/ruby-amqp/bunny/Bunny/Session#open%3F-instance_method).

miha-plesko replied:
@gberginc this seems to be the best solution, thank you.

miha-plesko commented:
I've added a drop-down for security option as @AparnaKarve suggested, even if I've only offered a non-ssl option for now. As far as I can tell, this works now 😄 

\cc @agrare

gberginc commented:
@miq-bot remove_label wip

gberginc commented:
@miha-plesko one last (I promise :wink:) comment from my side: please update the screenshot

gberginc commented on app/controllers/ems_common.rb:
```diff
@@ -813,6 +813,7 @@ def form_instance_vars
     @openstack_api_versions = retrieve_openstack_api_versions
     @vmware_cloud_api_versions = retrieve_vmware_cloud_api_versions
     @emstype_display = model.supported_types_and_descriptions_hash[@ems.emstype]
+    @vmware_cloud_amqp_security_protocols = retrieve_vmware_cloud_amqp_security_protocols
```
Since this is AMQP, unrelated to the underlying provider, do you think we need to have provider-specific lists?

gberginc commented on app/controllers/ems_common.rb:
```diff
@@ -847,6 +848,10 @@ def retrieve_openstack_amqp_security_protocols
   end
 
+  def retrieve_vmware_cloud_amqp_security_protocols
+    [[_('Non-SSL'), 'non-ssl']]
```
Same as above

gberginc commented on app/views/layouts/angular-bootstrap/_endpoints_angular.html.haml:
```diff
@@ -101,6 +101,17 @@
                        "prefix"                      => "#{prefix}",
                        "reset-validation-status"     => "#{prefix}_auth_status")
+    .col-md-8{"ng-if" => "emsCommonModel.emstype == 'vmware_cloud'"}
+      - security_protocols = @vmware_cloud_amqp_security_protocols
```
If you do choose to merge openstack and vmware amqp security protocols, fix it here as well.

gberginc replied:
As you are using OpenStack's JS handler in ng-change below, I think you have to use the same thing!

gberginc commented on app/views/layouts/angular/_multi_auth_credentials.html.haml:
```diff
@@ -188,13 +188,59 @@
                                     :basic_info_needed => true}
+          .col-md-12{"ng-if" => "#{ng_model}" == "emsCommonModel" && "#{ng_model}.emstype == 'vmware_cloud'"}
+            %div{"ng-if" => "emsCommonModel.event_stream_selection != 'amqp'"}
+              %div{"ng-init" => "emsCommonModel.event_stream_selection = 'none'"}
```
Why is this different than in the case of OpenStack (L154-155  -- there are no additional divs)?

gberginc replied:
I see that it works, I just wonder if it would be possible to make it resemble the OpenStack completely?

gberginc commented on app/views/layouts/angular/_multi_auth_credentials.html.haml:
```diff
@@ -188,13 +188,59 @@
               %div{"ng-if" => "emsCommonModel.event_stream_selection == 'amqp'"}
                 = _("Used to authenticate with OpenStack AMQP Messaging Bus for event handling. Configure AMQP if eventing is not enabled on Ceilometer.")
               %div{"ng-if" => "emsCommonModel.event_stream_selection == 'ceilometer'"}
                 = _("Select Ceilometer if eventing is not enabled on AMQP.")
```
However, here is the div that might be the one that you are creating above? Can you replicate this in VMware below?

miha-plesko commented:
@gberginc I've refactored the code in _multi_auth file so that it reuses as much OpenStack code as possible. Thank you for insisting on doing this, it really is better that way. Please check if there are some further modifications needed for this PR to be merged.

gberginc commented:
Thanks @miha-plesko for these last changes!

LGTM 👍

AparnaKarve commented:
@miha-plesko @gberginc Found an issue while testing in the UI, after which I pretty much stopped testing.
When I switch from `VMWare VCloud` to `Openstack` and then back to `VMWare VCloud`, the `default` and `amqp` tabs vanish. Likewise, the tabs on `Openstack` also vanish when I switch back to `Openstack`.

<img width="1267" alt="screen shot 2016-10-20 at 10 06 57 am" src="https://cloud.githubusercontent.com/assets/1538216/19570046/0da49258-96ad-11e6-8b60-fd6ce5dff4e7.png">

Please exercise caution with those `ng-if`s

Also, I did not see `Security Protocol` displayed when I chose `VMWare VCloud`

Is `VMWare VCloud` supposed to work exactly like `Openstack`? 
If yes, then we do not need those many changes in the hamls
If not, can you give me the exact difference in behavior for the 2 providers? (Maybe we could work up a better logic)

gberginc commented:
Thank you and apologies @AparnaKarve. Not sure what went wrong but just before the last push, we were switching between openstack and vmware 😞. We'll fix this asap.

Regarding the no `Security Protocol` -- are you referring to the `Default` or `Events` tab? The former does not have this currently, but the latter has it.

I don't quite understand your comment regarding "too many changes". This PR is only addressing the AMQP events tab and the changes I see there are mainly done to reuse as much code as possible. Unfortunately, VMware does not support ceilometer so we added the "none" option to allow users to switch between using AMQP or not using events at all.

We are open for suggestions.

@miha-plesko what if you split this into two PRs, one for the changes in the provider and the other for the UI?

gberginc commented:
@AparnaKarve https://github.com/ManageIQ/manageiq/pull/11682/commits/13f5a5ac517818d48da557f44ada55b3928043af shows what went wrong 😞. Sorry again for making you go through a review without us actually checking that the last little change was done properly!

AparnaKarve commented:
@gberginc No worries. I will retest this shortly.

AparnaKarve commented:
@gberginc Quick update: The above issue is now resolved, so that's great 👍 
Let me go through the haml code changes in detail and get back to you.
Thanks.

AparnaKarve commented:
@gberginc I've reviewed the UI changes. Other than that one minor change for i18n, everything else looks great!

miha-plesko commented:
Thank you for testing, @AparnaKarve. What's with the i18n that you've mentioned earlier? Would you like me to update some texts?

AparnaKarve commented:
@miha-plesko Whenever you introduce a new text in the UI, all you need to do is make sure that it is called like this `_("Some Text")` so that it gets extracted (by a rake task) and then translated by the translators.
As I mentioned above in the Review changes, you need to change `None` to `_("None")` and that should take care of i18n for that string.

You dont have to worry about `AMQP` since it is a standard abbreviation, or even `Ceilometer` for that matter since it's an Openstack terminology, meaning we don't need to translate those words. @mzazrivec would that be correct about abbreviations and standard terminology words?

miha-plesko commented:
Yesterday was the longest downtime of GitHub I've ever experienced --> 6 hours!

@AparnaKarve thank you for explaining about GUI translations, I've wrapped None into _("None"). Please let me know if there are any further modifications needed for the GUI.

I have split this PR into two commits: one for fine-tuning backend and one for implementing the frontend. When I was testing GUI I've discovered a hidden bug with backend, therefore I've added some new changes to the backend commit. But I believe these two commits belong to the same PR since it is difficult to test one without another. Will this work for you guys?

Ofc, should you insist on breaking this into two PRs I'm fine with it as well 😄

gberginc commented on app/models/manageiq/providers/vmware/manager_auth_mixin.rb:
```diff
@@ -17,7 +18,14 @@ def verify_credentials(auth_type = nil, options = {})
           _log.error("Error Class=#{err.class.name}, Message=#{err.message}")
           raise miq_exception
         end
-      when 'amqp' then verify_amqp_credentials(options)
```
Now that the two verifications are almost identical, I wonder if you could change the order of the code, i.e.

```
being
  case auth_type.to_s
    when 'default' then
      ..
    when 'amqp' then 
      ..
    else
      ..
  end
rescue ...
```

miha-plesko replied:
Ahhh, it reads better now. Thank you for suggesting this!
The `else` part error message will get prefixed with "Unexpected response returned from system: #{err.message}", which sounds reasonable.

gberginc commented on app/models/manageiq/providers/vmware/manager_auth_mixin.rb:
```diff
@@ -1,5 +1,6 @@
 module ManageIQ::Providers::Vmware::ManagerAuthMixin
   extend ActiveSupport::Concern
+  require 'fog/vcloud_director'
```
Move this to the beginning of the file?

miha-plesko replied:
Done. I put it here to make it similar to other `require_nested`, but it makes more sense to put it above.

gberginc commented on app/models/manageiq/providers/vmware/manager_auth_mixin.rb:
```diff
@@ -36,10 +44,23 @@ def connect(options = {})
+        MiqException::MiqUnreachableError.new "Login attempt timed out"
+      when Excon::Errors::SocketError
+        MiqException::MiqHostError.new "Socket error: #{err.message}"
+      when MiqException::MiqInvalidCredentialsError, MiqException::MiqHostError # from amqp
```
Are these really only possible from the AMQP? The error names indicate they may be more generic, so in this case I would suggest to remove the `from amqp` comment.

miha-plesko replied:
Yes, they can only come from AMQP since DEFAULT raises Fog exceptions directly (which is actually the very reason why `require 'fog/vcloud_director'` at the top of the file is needed). But I've removed the comment anyway since in the future perhaps METERING will wrapp its errors in this generic form as well.

agrare commented:
@gberginc provider side looks good

gberginc commented:
@AparnaKarve do we have a go on this one? @miha-plesko has added logic into the provider which is now even more solid. He also made the changes necessary for the i18n as you requested in your last review.

gberginc commented:
@miha-plesko please fix the case/when indentation and "&& instead of and". Other rubocop comments I believe are ok as you were reusing existing code.

AparnaKarve commented on app/assets/javascripts/controllers/ems_common/ems_common_form_controller.js:
```diff
@@ -327,6 +327,7 @@ ManageIQ.angular.app.controller('emsCommonFormController', ['$http', '$scope', '
       $scope.emsCommonModel.default_api_port = "8443";
     } else if ($scope.emsCommonModel.emstype === 'vmware_cloud') {
       $scope.emsCommonModel.default_api_port = "443";
+      $scope.emsCommonModel.event_stream_selection = "none";
```
@miha-plesko Can I request for one more change in the UI? (I fixed something similar for one other provider, and wanted to make sure we address that here as well)
Right here, can we add --
`$scope.emsCommonModel.amqp_security_protocol = 'non-ssl';`

The Validate button lights up even if no Security Protocol selection is made. So in order to avoid that, a default Security value here will help. The above change will ensure that the `Non-SSL` protocol is always selected for AMQP.

miha-plesko replied:
I've actually already spent some time on getting this chosen by default but gave up eventually. I was trying to go with `selected` option:
```javascript
options_for_select([...], {selected: ..., disabled: ...})
```
 in this file [here](https://github.com/ManageIQ/manageiq/blob/master/app/views/layouts/angular-bootstrap/_endpoints_angular.html.haml#L86), but that didn't work so I gave up.
Anyway, thanks for helping me out 😄

AparnaKarve commented:
@gberginc This is getting really close. Once we have the above change that I suggested, this should be good to go. Thanks.

miha-plesko commented:
@AparnaKarve @gberginc thank you for the review, I've applied the changes and repushed. I have a feeling that we're converging towards excelence here and that is awesome 😄 Is there anything we could do to make it even better, I'd be glad to do it.

AparnaKarve commented:
@miha-plesko Thanks!

@h-kataria This is LGTM 👍 . Can you please get this merged?

sergio-ocon commented:
:astonished:

h-kataria commented:
@miha-plesko will merge this PR once conflicts are resolved.

miha-plesko commented:
@h-kataria Rebased

sergio-ocon commented:
Hey @gtanzillo @agrare @miha-plesko. Should not be backported to Euwe? We need this functionality to get proper support of the provider?

gberginc commented:
@sergio-ocon True. The provider will work and update in regular intervals, but for using the events this is mandatory. #11362, a backend support for this, was also backported.

sergio-ocon commented:
@gberginc Thanks.
@miq-bot add_label euwe/yes

chessbyte commented:
Euwe Backport details:
```sh
$ git log -1
commit f2229fe9213a1de7189c7676b83b7266cedd8fcd
Author: Harpreet Kataria <hkataria@redhat.com>
Date:   Thu Oct 27 18:30:30 2016 -0400

    Merge pull request #11682 from xlab-si/add_vmware_vcloud_event_catcher_gui
    
    Add GUI with inputs for VMware vCloud event monitoring
    (cherry picked from commit b7bb8d13281d54921ee834a6a4a0c5837bc7cde0)
```
